### PR TITLE
feat: Add maxSwapAmount.

### PR DESCRIPTION
**Pull request by johngrantuk**

### CODE REVIEW

File: typescript/src/weighted/weightedPool.ts [hunk 1]
```diff
@@ -8,13 +14,28 @@ import {
...
+    public _MAX_IN_RATIO = 300000000000000000n;
+    public _MAX_OUT_RATIO = 300000000000000000n;
 
     constructor(poolState: {
         weights: bigint[];
     }) {
         this.normalizedWeights = poolState.weights;
     }
 
+    getMaxSwapAmount(swapParams: MaxSwapParams): bigint {
```
Comment (brunoguerios) [inline typescript/src/weighted/weightedPool.ts:27]: I'm not really sure the best way to organize things, but it feels like we're breaking a pattern here.
If we decide to keep all math within WeightedMath, then we could have a `computeMaxSwapAmount` method that can reuse `_MAX_IN_RATIO` already set there.
What do you think?

**Review by brunoguerios — APPROVED** [review]
LGTM ✅ 

I left a suggestion to restructure it a bit, but non-critical. Feel free to organize as you prefer.

About the feature itself, I know it introduces a new requirement for pool devs that want to include their pools to Balancer Maths, but it should be ok, right?
I mean, that's something you'd expect a pool developer to think a bit before actually even creating a new pool type.
Maybe it's an opportunity to remind them that it's something worth spending some time on 😅

File: typescript/src/weighted/weightedPool.ts (hunk 1 above)
Comment (johngrantuk) [reply typescript/src/weighted/weightedPool.ts:26]: I forgot we already had the MAX values defined in maths! I've use them as an import instead.
Lets keep `PoolMath` as close to the contracts as possible. The maxSwap calculation is kind of a custom requirement for our SOR use case (more so for non-weighted pools).

**Comment by johngrantuk** [comment]
> LGTM ✅
> 
> I left a suggestion to restructure it a bit, but non-critical. Feel free to organize as you prefer.
> 
> About the feature itself, I know it introduces a new requirement for pool devs that want to include their pools to Balancer Maths, but it should be ok, right? I mean, that's something you'd expect a pool developer to think a bit before actually even creating a new pool type. Maybe it's an opportunity to remind them that it's something worth spending some time on 😅

I kind of felt the same about adding it here instead of the API directly but I can already think of one place we'd like to reuse it, Paraswap integration, so I think its best here.

I'm not sure all pool developers do think about it up front without a nudge so it probably is a good opportunity!

**Pull request merged**